### PR TITLE
includePayload feature (fixes issue #3272)

### PR DESCRIPTION
**Pull request by Anjali-NEC**
Fixed issue #3272

### CODE REVIEW

**Comment by Anjali-NEC** [comment]
Hi @fgalan @kzangeli sir,

I've fixed the issue #3272. Please check PR and if it is ok then please merge the PR in master branch.

**Comment by fgalan** [comment]
Thanks for your contribution!

Two main comments:

1. You should include functional tests (.test files) to cover the new functionality. Some information about how to create .test files here (maybe slightly obsolete, but mostly useful): https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/devel/cookbook.md#adding-a-functional-test-case
2. You should describe the new field in subscriptions payload as an "Orion extension" of the NGSIv2 spec. I mean, in the following file, in a new section in https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/user/ngsiv2_implementation_notes.md named "Additional fields in `httpInfo` for custom notifications"

Thanks!

File: doc/manuals/user/ngsiv2_implementation_notes.md [hunk 1]
```diff
@@ -238,6 +238,8 @@ The particular validations that Orion implements on NGSIv2 subscription payloads
 * **throttling**: optional (must be an integer)
 * **expires**: optional (must be a date or empty string "")
 * **status**: optional (must be a valid status keyword)
+* **Orion extension**:
```
Comment (fgalan) [inline doc/manuals/user/ngsiv2_implementation_notes.md:241]: Maybe I didn't explain clearly enough...

What I mean is to create a new section in this .md document, I mean, this way:

```
## Additional fields in `httpInfo` for custom notifications
```

and inside that section explain the new field `includePayload`.

The section `## Subscription payload validations` doesn't need any change. Please, rollback your addition there so it looks as before.

File: test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+--SHELL--
+
+#
+# 01. Create E1/T1
+# 02. Create custom subscription with empty PAYLOAD and URL
+# 03. Create custom subscription with empty PAYLOAD, URL and VERB
+# 04. Create custom subscription with non-empty PAYLOAD, URL, VERB
+# 05. Create custom subscription with empty PAYLOAD, URL and VERB
+# 06. Dump accumulator to see the notification
```
Comment (fgalan) [inline test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test:39]: None of this steps is actually testing the functionality in this PR, i.e. the `includePayload` field. Thus, I'd suggest the following script of steps:

* 01. Create E1/T1
* 02. Create custom subscription with payload="", includePayload=false
* 03. Create custom subscription with payload="", includePayload=true
* 04. Create custom subscription without payload, includePayload=false
* 05. Create custom subscription without paylaod, includePayload=true
* 06. Create custom subscription without payload="foo", includePayload=false
* 07. Create custom subscription without paylaod="foo", includePayload=true
* 08. Dump accumulator to see the notification
```

Note that VERB is no related with this functionality so it is not needed to change it for the tests. URL is needed always (so there is no need of mentioning it in the step names)

Comment (Anjali-NEC) [reply test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test:39]: Hi @fgalan  sir,

Thank you for your guidance. I have one query redarding test case:
> `# 06. Create custom subscription without payload="foo", includePayload=false`
> `# 07. Create custom subscription without paylaod="foo", includePayload=true`

I think this should be:
> `# 06. Create custom subscription with payload="foo", includePayload=false`
> `# 07. Create custom subscription with paylaod="foo", includePayload=true`

if i have created custom subscription with payload="foo" I am getting the response:

POST http://localhost:REGEX(\d+)/notify
Fiware-Servicepath: /
Content-Length: 3
User-Agent: orion/REGEX(\d+\.\d+\.\d+.*)
Ngsiv2-Attrsformat: custom
Host: localhost:REGEX(\d+)
Accept: application/json
Content-Type: text/plain; charset=utf-8
Fiware-Correlator: REGEX([0-9a-f\-]{36})

No JSON object could be decoded=======================================

File: doc/manuals/user/ngsiv2_implementation_notes.md (hunk 1 above)
Comment (Anjali-NEC) [reply doc/manuals/user/ngsiv2_implementation_notes.md]: Updated

File: test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test (hunk 2 above)
Comment (fgalan) [reply test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test:39]: You are right in both cases :)

First one is just a typo.

Second one is due to the accumulator-server.py used by the test tries to decode the payload as a JSON. I don't remember right now if it has some option to process plain text (maybe it has...). However, to keep it simple, maybe we can do:

```
    # 06. Create custom subscription with payload="{ %22x%22: 1}", includePayload=false
    # 07. Create custom subscription with paylaod="{ %22x%22: 1}", includePayload=true
```

%22 is to scape quotes (see https://fiware-orion.readthedocs.io/en/master/user/forbidden_characters/index.html#custom-payload-special-treatment)

File: doc/manuals/user/ngsiv2_implementation_notes.md [hunk 3]
```diff
@@ -455,3 +456,11 @@ which attributes to include in the response to the query) and unary attribute fi
 `expression` (to return only entities which have these attributes). Use them instead.
 
 [Top](#top)
+
+## Additional fields in `httpInfo` for custom notifications
+
+* includePayload: Additional field in `httpInfo` for custom notifications to check not including payload or 0-length requests.
+set 'true' in the case of not including payload in the request.
```
Comment (fgalan) [inline doc/manuals/user/ngsiv2_implementation_notes.md:463]:
```suggestion
Orion supports a boolean field `includePayload` in `httpInfo` for custom notifications that can be used to avoid payload in the outgoing notification. If `includePayload` is set to `true` then no payload is included in the notification (and `payload` field, if used, is ignored).

Default value if `includePayload` is omitted is `false`.
```

Comment (Anjali-NEC) [reply doc/manuals/user/ngsiv2_implementation_notes.md]: Updated

File: test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test (hunk 2 above)
Comment (Anjali-NEC) [reply test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test]: Hi @fgalan sir,
Thank you for your suggestion. I have updated test cases as per your comment.

File: test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+HTTP/1.1 201 Created
+Content-Length: 0
+Location: /v2/subscriptions/REGEX([0-9a-f]{24})
+Fiware-Correlator: REGEX([0-9a-f\-]{36})
+Date: REGEX(.*)
+
+
+
+08. Dump accumulator to see the notification
+============================================
```
Comment (fgalan) [inline test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test:289]: Note that all the 6 notifications include payload. This is not correct, the cases correesponding to includePayload=false shouldn't include any payload (content-length should be 0 in that cases).

File: src/lib/apiTypesV2/HttpInfo.cpp [hunk 5]
```diff
@@ -116,6 +116,7 @@ void HttpInfo::fill(const BSONObj& bo)
   if (this->custom)
   {
     this->payload  = bo.hasField(CSUB_PAYLOAD)? getStringFieldF(bo, CSUB_PAYLOAD) : "";
+    this->includePayload = this->payload == ""? true : false;
```
Comment (fgalan) [inline src/lib/apiTypesV2/HttpInfo.cpp:119]: I have realized on this...

As far as I understand, you are setting the value of includePayload based on another field (payload). As far as I understand, it should be an idependent parametner of the subscription, managed the same way any other CSUB_xxxx in the database.

Have a look to CSUB_ONLYCHANGED (another boolean in the subs information). It should be managed in a similar way.

File: CHANGES_NEXT_RELEASE [hunk 6]
```diff
@@ -1,3 +1,4 @@
+- Fix: Custom notification: sending 0-length request #3272
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE:1]:
```suggestion
- Fix: notification.includePayload to avoid sending payloads in custom notification (#3272)
```

File: test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+#
+# 01. Create E1/T1
+# 02. Create custom subscription with payload="", includePayload=false
+# 03. Create custom subscription with payload="", includePayload=true
+# 04. Create custom subscription without payload, includePayload=false
+# 05. Create custom subscription without paylaod, includePayload=true
+# 06. Create custom subscription with payload="{ %22x%22: 1}", includePayload=false
+# 07. Create custom subscription with paylaod="{ %22x%22: 1}", includePayload=true
+# 08. Dump accumulator to see the notifications
+#
```
Comment (fgalan) [inline test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test:42]: I think it would be a good idea to add:

```
# 09. GET /v2/subscriptions to check subscriptions payload
```

to check that the payload of the subscriptions is as expected.

Comment (Anjali-NEC) [reply test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test:42]: Updated

File: test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test (hunk 4 above)
Comment (Anjali-NEC) [reply test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test:298]: Updated

File: src/lib/apiTypesV2/HttpInfo.cpp (hunk 5 above)
Comment (Anjali-NEC) [reply src/lib/apiTypesV2/HttpInfo.cpp]: Updated

File: doc/manuals/user/ngsiv2_implementation_notes.md [hunk 8]
```diff
@@ -455,3 +456,10 @@ which attributes to include in the response to the query) and unary attribute fi
 `expression` (to return only entities which have these attributes). Use them instead.
 
 [Top](#top)
+
+## Additional fields in `httpInfo` for custom notifications
+
+Orion supports a boolean field `includePayload` in `httpInfo` for custom notifications that can be used to avoid payload in the outgoing notification. If `includePayload` is set to `true` then no payload is included in the notification (and `payload` field, if used, is ignored).
+Default value if `includePayload` is omitted is `false`.
```
Comment (fgalan) [inline doc/manuals/user/ngsiv2_implementation_notes.md:463]:
```suggestion
Default value if `includePayload` if omitted is `false`.
```

Comment (Anjali-NEC) [reply doc/manuals/user/ngsiv2_implementation_notes.md]: Updated

File: test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test [hunk 9]
```diff
@@ -0,0 +1,567 @@
...
+            "type": "T1"
+        }
+    ],
+    "subscriptionId": "REGEX([0-9a-f]{24})"
+}
+=======================================
+
+
+09. GET the subscription
+========================
```
Comment (fgalan) [inline test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test:391]: Note that includedPayload field is not included in the JSON of the subscriptions in this step and it must be there. It should behave similar to other optional boolean fields in subscription, such as "onlyChanged".

Let me illustrate with an example that you can reproduce yourself. We create three subscriptions, each one with a different possibilities for onlyChanged: not included, set to true and set to false.

```
curl -v localhost:1026/v2/subscriptions -s -S -H 'Content-Type: application/json' -d @- <<EOF
{
  "description": "Without onlyChanges field",
  "subject": {
    "entities": [
      {
        "id": "Room1",
        "type": "Room"
      }
    ]
  },
  "notification": {
    "http": {
      "url": "http://localhost:1028/notify"
    }  
  }
}
EOF

curl -v localhost:1026/v2/subscriptions -s -S -H 'Content-Type: application/json' -d @- <<EOF
{
  "description": "With onlyChanges set to true",
  "subject": {
    "entities": [
      {
        "id": "Room1",
        "type": "Room"
      }
    ]
  },
  "notification": {
    "http": {
      "url": "http://localhost:1028/notify"
    },
    "onlyChangedAttrs": true
  }
}
EOF

curl -v localhost:1026/v2/subscriptions -s -S -H 'Content-Type: application/json' -d @- <<EOF
{
  "description": "With onlyChanges set to false",
  "subject": {
    "entities": [
      {
        "id": "Room1",
        "type": "Room"
      }
    ]
  },
  "notification": {
    "http": {
      "url": "http://localhost:1028/notify"
    },
    "onlyChangedAttrs": false
  }
}
EOF
```

now we get them and get the following:

```
$ curl -s -S  localhost:1026/v2/subscriptions | python -mjson.tool

[
    {
        "description": "Without onlyChanges field",
        "id": "5f7ee8b84f85b328ecc958ff",
        "notification": {
            "attrs": [],
            "attrsFormat": "normalized",
            "http": {
                "url": "http://localhost:1028/notify"
            },
            "onlyChangedAttrs": false
        },
        "status": "active",
        "subject": {
            "condition": {
                "attrs": []
            },
            "entities": [
                {
                    "id": "Room1",
                    "type": "Room"
                }
            ]
        }
    },
    {
        "description": "With onlyChanges set to true",
        "id": "5f7ee8b84f85b328ecc95900",
        "notification": {
            "attrs": [],
            "attrsFormat": "normalized",
            "http": {
                "url": "http://localhost:1028/notify"
            },
            "onlyChangedAttrs": true
        },
        "status": "active",
        "subject": {
            "condition": {
                "attrs": []
            },
            "entities": [
                {
                    "id": "Room1",
                    "type": "Room"
                }
            ]
        }
    },
    {
        "description": "With onlyChanges set to false",
        "id": "5f7ee8b94f85b328ecc95901",
        "notification": {
            "attrs": [],
            "attrsFormat": "normalized",
            "http": {
                "url": "http://localhost:1028/notify"
            },
            "onlyChangedAttrs": false
        },
        "status": "active",
        "subject": {
            "condition": {
                "attrs": []
            },
            "entities": [
                {
                    "id": "Room1",
                    "type": "Room"
                }
            ]
        }
    }
]
```

In summary:

* `"onlyChange": false` for the cases of ommitng and set to false
* `"onlyChange": true` for the case of set to true

The new field includePayload should behave similarly (with the difference that, in this case, omitting will result in true instead of false).

File: doc/manuals/user/ngsiv2_implementation_notes.md [hunk 10]
```diff
@@ -455,3 +456,10 @@ which attributes to include in the response to the query) and unary attribute fi
 `expression` (to return only entities which have these attributes). Use them instead.
 
 [Top](#top)
+
+## Additional fields in `httpInfo` for custom notifications
+
+Orion supports a boolean field `includePayload` in `httpInfo` for custom notifications that can be used to avoid payload in the outgoing notification. If `includePayload` is set to `true` then no payload is included in the notification (and `payload` field, if used, is ignored).
+Default value if `includePayload` if omitted is `false`.
```
Comment (fgalan) [inline doc/manuals/user/ngsiv2_implementation_notes.md:463]: I think that for backward compatibility default value should be `true` (as currently in Orion, previously to this PR, the payload is always included).

Sorry for not realizing of this until now...

File: src/lib/apiTypesV2/HttpInfo.h [hunk 11]
```diff
@@ -47,6 +47,7 @@ struct HttpInfo
   std::map<std::string, std::string>  headers;
   std::string                         payload;
   bool                                custom;
+  bool                                includePayload;
```
Comment (fgalan) [inline src/lib/apiTypesV2/HttpInfo.h:50]: Class constructors in HttpInfo.cpp should be modified to ensure that includePayload is initialized with the right default value (true).

File: CHANGES_NEXT_RELEASE [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+- Fix: POST /v2/registrations allows forbidden attribute names in NGSIv2 (#3468)
+- Fix: log WARN about insecureNotifications mode (#3013)
+- Fix: avoid creating orion database in -multiservice mode (#3179)
+- Fix: upgrade rapidjson dependecy from 1.0.2 to 1.1.0
+- Fix: upgrade microhttpd dependency from 0.9.48 to 0.9.70
+- Fix: potencial race condition related with DateTime attributes due to the usage of thread unsafe functions
+- Fix: wrong 204 code was responded (instead of 501) when using idPattern in POST /v2/op/update operation (#3427)
+- Fix: changed the name of pretty much all the environment variables for Orion. E.g. adding prefix 'ORION_'
+- Add: milliseconds support in DateTime attributes and metadata, included dateCreated and dateModified built-ins (#432, #2670, #3412, #3666)
+
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE:12]:
```suggestion
- Fix: notification.includePayload to avoid sending payloads in custom notification (#3272)
```

Note we have done a release while your PR was opened. The change of release flushes the CHANGES_NEXT_RELEASE file so, after that, only the line corresponding to your PR should appear.

Comment (Anjali-NEC) [reply CHANGES_NEXT_RELEASE]: Updated

File: doc/manuals/user/ngsiv2_implementation_notes.md [hunk 13]
```diff
@@ -455,3 +456,10 @@ which attributes to include in the response to the query) and unary attribute fi
 `expression` (to return only entities which have these attributes). Use them instead.
 
 [Top](#top)
+
+## Additional fields in `httpInfo` for custom notifications
+
+Orion supports a boolean field `includePayload` in `httpInfo` for custom notifications that can be used to avoid payload in the outgoing notification. If `includePayload` is set to `true` then no payload is included in the notification (and `payload` field, if used, is ignored).
+Default value if `includePayload` if omitted is `true`.
```
Comment (fgalan) [inline doc/manuals/user/ngsiv2_implementation_notes.md:463]:
```suggestion
Orion supports a boolean field `includePayload` in `httpInfo` for custom notifications that can be used to avoid payload in the outgoing notification. If `includePayload` is set to `false` then no payload is included in the notification (and `payload` field, if used, is ignored).
Default value if `includePayload` if omitted is `true` (that is, payload is included by default).
```

Note the default `true` will ensuse backward compatibility.

File: src/lib/apiTypesV2/HttpInfo.cpp [hunk 14]
```diff
@@ -78,7 +78,7 @@ std::string HttpInfo::toJson()
 
   if (custom)
   {
-    if (!this->payload.empty())
+    if (!this->includePayload)
```
Comment (fgalan) [inline src/lib/apiTypesV2/HttpInfo.cpp:81]: Not sure but I'd say this change needs to be reverted. Note this is a method to render part of the subscription payload (i.e. the one you get with `GET /v2/subscriptions`) and the `payload` field (if exist) has to be shown no matter the value of `includePayload`.

File: src/lib/apiTypesV2/Subscription.cpp [hunk 15]
```diff
@@ -146,6 +146,15 @@ std::string Notification::toJson(const std::string& attrsFormat)
...
 
+  if (this->httpInfo.payload.empty() && this->includePayload)
+  {
+    jh.addBool("includePayload", true);
+  }
+  else
+  {
+    jh.addBool("includePayload", false);
+  }
+
```
Comment (fgalan) [inline src/lib/apiTypesV2/Subscription.cpp:157]: This fragment should be moved to HttpInfo::toJson(), excluding `this->httpInfo.payload.empty()` from the if condition. Note that the new field shouldd be inside `httpCustom` but not in `notification` directly.

File: src/lib/mongoBackend/dbConstants.h [hunk 16]
```diff
@@ -110,6 +110,7 @@
 #define CSUB_HEADERS                 "headers"
 #define CSUB_QS                      "qs"
 #define CSUB_PAYLOAD                 "payload"
+#define CSUB_INCLUDEPAYLOAD          "includePayload"
```
Comment (fgalan) [inline src/lib/mongoBackend/dbConstants.h:113]: New field should be added to database model documentation. In particular to this section:

https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/admin/database_model.md#csubs-collection

File: src/lib/ngsiNotify/Notifier.cpp [hunk 17]
```diff
@@ -238,7 +238,7 @@ static std::vector<SenderThreadParams*>* buildSenderParamsCustom
     //
     // 3. Payload
     //
-    if (httpInfo.payload.empty())
+    if (httpInfo.includePayload)
```
Comment (fgalan) [inline src/lib/ngsiNotify/Notifier.cpp:241]: Not sure about this fix. I understand the idea in the original code is something like this:

```
if (http.payload.empty())
{
   <not use any special payload and use the NGSI default one>
}
else
{
   <use the payload from the custom notification>
}
```

but note that in both branches of the if a payload is used in the notification.

Thus, the check for includePayload should be a "higher level". Something like this:

```
if (httpInfo.includePayload)
{
   if (http.payload.empty())
   {
      <not use any special payload and use the NGSI default one>
   }
   else
   {
      <use the payload from the custom notification>
   }
}
else
{
   <not include paylaod at all in the notification (in fact, maybe this else block is not needed at all if there is nothing to do in this case)>
}
```

However, I don't like too much the double-if as it makes readibility worse, from my point of view. Maybe you can find a way of refactoring to improve my suggestion :)

File: test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test [hunk 18]
```diff
@@ -0,0 +1,567 @@
+# Copyright 2020 Telefonica Investigacion y Desarrollo, S.A.U
```
Comment (fgalan) [inline test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test:1]: I'd suggest to split this test in two .test, one focused in the subscriptions API and another in notifications.

The one focused in subscriptions API could be as follows:

```
# 01. Create sub1 with payload="abc", includePayload=false
# 02. Create sub2 with payload="", includePayload=false
# 03. Create sub3 without payload, includePayload=false
# 04. Create sub4 with payload="abc", includePayload=true
# 05. Create sub5 with payload="", includePayload=true
# 06. Create sub6 without payload, includePayload=true
# 07. Create sub7 with payload="abc", without includePayload
# 08. Create sub8 with payload="", without includePayload
# 09. Create sub9 without payload, without includePayload
# 10. GET /v2/subscription and see the 9 subscriptions
```

The one focused in notifications could be as follows:

```
01. Create sub1 with payload="abc", includePayload=false
02. Create sub2 without payload, includePayload=false
03. Create entity E1/T1
04. Dumps acummulator (see 2 notifications, both without payload)
```

Comment (fgalan) [reply test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test:1]: With regards the first case (subscription API test) please take into account the considerations already described in https://github.com/telefonicaid/fiware-orion/pull/3683/files#r501631288

File: doc/manuals/user/ngsiv2_implementation_notes.md [hunk 19]
```diff
@@ -23,6 +23,7 @@
 * [Registrations](#registrations)
 * [`keyValues` not supported in `POST /v2/op/notify`](#keyvalues-not-supported-in-post-v2opnotify)
 * [Deprecated features](#deprecated-features)
+* [Additional fields in `httpInfo` for custom notifications](#additional-fields-in-`httpInfo`-for-custom-notifications)
```
Comment (fgalan) [inline doc/manuals/user/ngsiv2_implementation_notes.md:26]:
```suggestion
* [Additional fields in `httpInfo` for custom notifications](#additional-fields-in-httpInfo-for-custom-notifications)
```

Otherwise it seems the link doesn't work.

**Comment by fgalan** [comment]
Thanks for your contribution!

I have done a new review of the PR. Please find my comments inline with the code.

In addition note that there is an (esasy to solve) conflict in CHANGES_NEXT_RELEASE. Please fix it, so travis can run again.

File: doc/manuals/user/ngsiv2_implementation_notes.md (hunk 19 above)
Comment (Anjali-NEC) [reply doc/manuals/user/ngsiv2_implementation_notes.md]: Updated

File: test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test (hunk 18 above)
Comment (Anjali-NEC) [reply test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test:1]: Added test case

File: src/lib/ngsiNotify/Notifier.cpp (hunk 17 above)
Comment (Anjali-NEC) [reply src/lib/ngsiNotify/Notifier.cpp:241]: updated

File: src/lib/mongoBackend/dbConstants.h (hunk 16 above)
Comment (Anjali-NEC) [reply src/lib/mongoBackend/dbConstants.h:113]: Added

File: src/lib/apiTypesV2/Subscription.cpp (hunk 15 above)
Comment (Anjali-NEC) [reply src/lib/apiTypesV2/Subscription.cpp]: Updated

File: src/lib/apiTypesV2/HttpInfo.cpp (hunk 14 above)
Comment (Anjali-NEC) [reply src/lib/apiTypesV2/HttpInfo.cpp]: Updated

File: doc/manuals/user/ngsiv2_implementation_notes.md (hunk 13 above)
Comment (Anjali-NEC) [reply doc/manuals/user/ngsiv2_implementation_notes.md]: Updated

File: CHANGES_NEXT_RELEASE [hunk 20]
```diff
@@ -1,6 +1,10 @@
...
+- Fix: notification.includePayload to avoid sending payloads in custom notification (#3272)
+- Add: -disableFileLog CLI parameter (ORION_DISABLE_FILE_LOG env var) to prevent Orion from logging into a file
+- Add: support 'true' and 'false' literals (in addition to 'TRUE' and 'FALSE') in flag-like env vars
+- Upgrade Dockerfile base image from centos7.7.1908 to centos7.9.2009
+- FIX BUG: Create normal files as 644 permissions for assure backup tools take configuration files
+- Add: -disableFileLog CLI parameter (ORION_DISABLE_FILE_LOG env var) to prevent Orion from logging into a file
+- Add: support 'true' and 'false' literals (in addition to 'TRUE' and 'FALSE') in flag-like env vars
+- Fix: crash on MHD_REQUEST_TERMINATED_CLIENT_ABORT situations (#3738)
+- Upgrade Dockerfile base image from centos7.7.1908 to centos7.9.2009
+- FIX BUG: Create normal files as 644 permissions for assure backup tools take configuration files
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE:10]: It seems a conflict in this file was not correctly solved. The file should show the current content in master + the lines corresponding to the changes in your PR.

I think that after accepting this suggestion it will be fixed:

```suggestion
- Add: -disableFileLog CLI parameter (ORION_DISABLE_FILE_LOG env var) to prevent Orion from logging into a file
- Add: support 'true' and 'false' literals (in addition to 'TRUE' and 'FALSE') in flag-like env vars
- Fix: notification.includePayload to configure sending payloads or not in custom notification (#3272)
- Fix: crash on MHD_REQUEST_TERMINATED_CLIENT_ABORT situations (#3738)
- Upgrade Dockerfile base image from centos7.7.1908 to centos7.9.2009
- FIX BUG: Create normal files as 644 permissions for assure backup tools take configuration files
```

**Comment by fgalan** [comment]
Looking to travis result (https://travis-ci.org/github/telefonicaid/fiware-orion/jobs/751152824) it seems that some functional test are still failing:

```
----------- Failing tests ------------------
  o  1960_failure_stats_in_db_and_notifications/failure_stats_in_db_and_notifications.test
  o  2015_notification_templates/custom_qs_and_headers_names.test
  o  2015_notification_templates/notification_templates.test
  o  2015_notification_templates/notification_templates_many_notifications.test
  o  2015_notification_templates/notification_templates_many_notifications_after_update.test
  o  2015_notification_templates/notification_templates_payload_as_xml.test
  o  2015_notification_templates/notification_templates_with_decoded_chars_in_notifications.test
  o  2207_not_spurious_decimals_in_custom_notifications/not_spurious_decimals_in_custom_notifications.test
  o  2280_custom_url_validity_check/custom_url_validity_check.test
  o  2893_case_insensitive_custom_notif_http_headers/case_insensitive_custom_notif_http_headers.test
  o  2893_case_insensitive_custom_notif_http_headers/case_insensitive_custom_notif_http_headers_with_text_plain.test
  o  2937_self_notification_protection/nonmodifiable_headers.test
  o  2937_self_notification_protection/self_notification_one_hop.test
  o  3154_notification_problem_with_custom_header/notification_problem_with_custom_header.test
```

File: CHANGES_NEXT_RELEASE (hunk 20 above)
Comment (fgalan) [reply CHANGES_NEXT_RELEASE]: Fixed in 2e3fef2

File: doc/manuals/admin/database_model.md [hunk 21]
```diff
@@ -321,6 +321,8 @@ Fields:
 -   **lastSuccessCode**: HTTP code (200, 400, 404, 500, etc.) returned by receiving endpoint last
     time a successful notification was sent.
     Not present if the subscription has never provoked a successful notification.
+-   **includePayload**: a boolean field that can be used to avoid payload in the outgoing notification. 
+    If `includePayload` is set to `false` then no payload is included in the notification (and `payload` field, if used, is ignored).
```
Comment (fgalan) [inline doc/manuals/admin/database_model.md:325]:
```suggestion
    If `includePayload` is set to `false` then no payload is included in the notification (and `payload` field, if used, is ignored). Default behaviour (if the field is not in the database) is `true` (i.e. to included payload).
```

File: src/lib/apiTypesV2/HttpInfo.cpp [hunk 22]
```diff
@@ -82,6 +82,14 @@ std::string HttpInfo::toJson()
...
       jh.addString("payload", this->payload);
     }
+    else if (this->includePayload)
+    {
+      jh.addBool("includePayload", false);
+    }
+    else
+    {
+      jh.addBool("includePayload", true);
+    }
```
Comment (fgalan) [inline src/lib/apiTypesV2/HttpInfo.cpp:92]: I'd say that it should be the other way around:

```suggestion
    else if (this->includePayload)
    {
      jh.addBool("includePayload", true);
    }
    else
    {
      jh.addBool("includePayload", false);
    }
```

I mean, if `this->includePayload` is `true` then you have to add `"includePayload": true` in the JSON payload. Or maybe I'm missing something? :)

File: src/lib/apiTypesV2/HttpInfo.h (hunk 11 above)
Comment (fgalan) [reply src/lib/apiTypesV2/HttpInfo.h:50]: Did you finally modify the constructor as described in comment above, please? Maybe I have missed it but I don't see the change in the PR...

File: src/lib/ngsiNotify/Notifier.cpp [hunk 23]
```diff
@@ -1,4 +1,4 @@
-﻿/*
+/*
```
Comment (fgalan) [inline src/lib/ngsiNotify/Notifier.cpp:1]: Where is the part of the code that depending on the value of includePayload includes the notification payload or not? I think that important piece is missing in the PR but, again, I could have missed something :)

File: test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test [hunk 24]
```diff
@@ -0,0 +1,187 @@
...
+=======================
+HTTP/1.1 201 Created
+Content-Length: 0
+Location: /v2/entities/E1?type=T1
+Fiware-Correlator: REGEX([0-9a-f\-]{36})
+Date: REGEX(.*)
+
+
+
+Dumps acummulator (see 2 notifications, both without payload)
```
Comment (fgalan) [inline test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test:141]: Note that both notifications in the test actually have payload... so it's not working as expected. See related comment: https://github.com/telefonicaid/fiware-orion/pull/3683/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.test#r576785475

File: test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test [hunk 25]
```diff
@@ -0,0 +1,187 @@
...
+  "type": "T1",
+  "A1": true
+}'
+orionCurl --url /v2/entities?options=keyValues --payload "$payload"
+echo
+echo
+
+
+echo "Dumps acummulator (see 2 notifications, both without payload)"
+echo "============================================================="
```
Comment (fgalan) [inline test/functionalTest/cases/3272_custom_notification_sending_0-length_request/custom_notification_sending_0-length_request.test:104]:
```suggestion
echo "04. Dumps acummulator (see 2 notifications, both without payload)"
echo "==============================================================="
```

REGEXPECT would need also adjustsments

**Comment by fgalan** [comment]
Instead of using a new field `includePayload` I'll try to re-work the PR in order to use the existing field `"payload": null` as originally suggested in https://github.com/telefonicaid/fiware-orion/issues/3272#issuecomment-664608406. I think that way is simpler from the point of view of the user (less parameters to use).

**Comment by fgalan** [comment]
Overpassed by PR https://github.com/telefonicaid/fiware-orion/pull/3850. Closing.
